File: app/routers/payroll.py
"""Payroll processing router."""
from fastapi import APIRouter, Depends, HTTPException
from app.models.payroll import PayrollProcess, PayrollFilter
from app.repositories import payroll_repo
from app.middleware.auth_middleware import get_current_user, require_role
from app.utils.helpers import paginate, pagination_meta

router = APIRouter(tags=["HR & Payroll"])


@router.post("/payroll/process", summary="Process monthly salaries for an employee")
def process_salary(body: PayrollProcess, user=Depends(require_role(["owner", "admin"]))):
    result, error = payroll_repo.process_salary(
        user_id=body.user_id, month=body.month,
        base_salary=body.base_salary, days_worked=body.days_worked,
        bonus=body.bonus, deductions=body.deductions,
        payment_method=body.payment_method, notes=body.notes or "",
    )
    if error:
        raise HTTPException(400, error)
    return {"ok": True, "data": result}


@router.get("/payroll", summary="List payroll records")
def list_payroll(
    month: str = None,
    user_id: int = None,
    page: int = 1,
    limit: int = 50,
    user=Depends(get_current_user),
):
    offset, limit, page = paginate(page, limit)
    rows, total = payroll_repo.list_payroll(month=month, user_id=user_id, offset=offset, limit=limit)
    return {"ok": True, "data": rows, "meta": pagination_meta(total, page, limit)}


@router.get("/payroll/summary/{month}", summary="Get overall payroll summary for a specific month")
def payroll_summary(month: str, user=Depends(get_current_user)):
    return {"ok": True, "data": payroll_repo.get_payroll_summary(month)}

File: frontend/src/app/api/payroll.ts
import { apiRawRequest } from "./client";

export type PayrollRow = {
  id: number;
  user_id: number;
  user_name?: string | null;
  month: string;
  base_salary: number;
  days_worked: number;
  bonus?: number;
  deductions?: number;
  net_salary?: number;
  payment_method?: string | null;
  notes?: string | null;
  created_at?: string;
};

export type PayrollProcessPayload = {
  user_id: number;
  month: string;
  base_salary: number;
  days_worked: number;
  bonus?: number;
  deductions?: number;
  payment_method?: string;
  notes?: string;
};

export type PayrollSummary = {
  total_records?: number;
  total_base_salary?: number;
  total_bonus?: number;
  total_deductions?: number;
  total_net_salary?: number;
};

export function getPayroll(token: string, month?: string) {
  const query = month ? `?month=${encodeURIComponent(month)}` : "";
  return apiRawRequest<{ ok: boolean; data: PayrollRow[]; meta?: unknown }>(`/payroll${query}`, { token });
}

export function processPayroll(token: string, payload: PayrollProcessPayload) {
  return apiRawRequest<{ ok: boolean; data: PayrollRow }>("/payroll/process", {
    method: "POST",
    token,
    body: payload
  });
}

export function getPayrollSummary(token: string, month: string) {
  return apiRawRequest<{ ok: boolean; data: PayrollSummary }>(`/payroll/summary/${month}`, { token });
}
